File: modules/05-tool/origin/tooluse-langchain.py
import asyncio
import getpass
import os

from langchain.agents import AgentExecutor, create_tool_calling_agent
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.tools import tool as langchain_tool
from langchain_google_genai import ChatGoogleGenerativeAI

# UNCOMMENT
# Prompt the user securely and set API keys as environment variables
# if needed.
os.environ["GOOGLE_API_KEY"] = getpass.getpass("Enter your Google API key: ")
os.environ["OPENAI_API_KEY"] = getpass.getpass("Enter your OpenAI API key: ")

try:
    # A model with function/tool calling capabilities is required.
    llm = ChatGoogleGenerativeAI(model="gemini-2.0-flash", temperature=0)
    print(f"✅ Language model initialized: {llm.model}")
except Exception as error:
    print(f"🛑 Error initializing language model: {error}")
    llm = None


# --- Define a Tool ---
@langchain_tool
def search_information(query: str) -> str:
    """Return simulated factual information for a few preset queries."""

    print(f"\n--- 🛠️ Tool Called: search_information with query: '{query}' ---")
    # Simulate a search tool with a dictionary of predefined results.
    simulated_results = {
        "weather in london": "The weather in London is currently cloudy with a temperature of 15°C.",
        "capital of france": "The capital of France is Paris.",
        "population of earth": "The estimated population of Earth is around 8 billion people.",
        "tallest mountain": "Mount Everest is the tallest mountain above sea level.",
        "default": "Simulated search result for '{query}': No specific information found, but the topic seems interesting.",
    }
    result = simulated_results.get(query.lower(), simulated_results["default"])
    print(f"--- TOOL RESULT: {result} ---")
    return result


tools = [search_information]

# --- Create a Tool-Calling Agent ---
if llm:
    # This prompt template requires an `agent_scratchpad` placeholder for
    # the agent's internal steps.
    agent_prompt = ChatPromptTemplate.from_messages(
        [
            ("system", "You are a helpful assistant."),
            ("human", "{input}"),
            ("placeholder", "{agent_scratchpad}"),
        ]
    )

    # Create the agent, binding the LLM, tools, and prompt together.
    agent = create_tool_calling_agent(llm, tools, agent_prompt)

    # AgentExecutor is the runtime that invokes the agent and executes
    # the chosen tools. The 'tools' argument is not needed here because
    # they are already bound to the agent.
    agent_executor = AgentExecutor(agent=agent, verbose=True, tools=tools)
else:
    agent_executor = None


async def run_agent_with_tool(query: str) -> None:
    """Invoke the agent executor with a query and print the final response."""

    if agent_executor is None:
        print("\n🛑 Agent is unavailable because LLM initialization failed.")
        return

    print(f"\n--- 🏃 Running Agent with Query: '{query}' ---")
    try:
        response = await agent_executor.ainvoke({"input": query})
        print("\n--- ✅ Final Agent Response ---")
        print(response["output"])
    except Exception as error:
        print(f"\n🛑 An error occurred during agent execution: {error}")


async def main() -> None:
    """Run sample queries for the LangChain tool-calling agent."""

    await run_agent_with_tool("What is the capital of France?")
    await run_agent_with_tool("What's the weather like in London?")
    await run_agent_with_tool("Tell me something about dogs.")


if __name__ == "__main__":
    asyncio.run(main())

File: modules/05-tool/typescript/tooluse-langchain.ts
import { ChatPromptTemplate } from "@langchain/core/prompts";
import {
  tool,
  type DynamicTool,
  type StructuredToolInterface,
} from "@langchain/core/tools";
import { ChatOpenAI } from "@langchain/openai";
import { AgentExecutor, createToolCallingAgent } from "langchain/agents";
import { resolve } from "node:path";
import { pathToFileURL } from "node:url";
import { z } from "zod";

type RuntimeConfig = {
  apiKey: string;
  baseUrl: string;
  model: string;
};

type AgentResponse = {
  output?: unknown;
};

type AgentExecutorLike = {
  invoke(input: { input: string }): Promise<AgentResponse>;
};

function resolveRuntimeConfig(
  env: NodeJS.ProcessEnv = process.env,
): RuntimeConfig {
  const apiKey: string = env.API_KEY?.trim() ?? "";
  if (apiKey.length === 0) {
    throw new Error("API_KEY is missing.");
  }

  return {
    apiKey,
    baseUrl: env.BASE_URL?.trim() || "https://api.openai.com/v1",
    model: env.MODEL?.trim() || "gpt-4o-mini",
  };
}

const searchInformationTool: DynamicTool<string> = tool(
  (query: string): string => {
    console.log(
      `\n--- Tool Called: search_information with query: '${query}' ---`,
    );

    const simulatedResults: Record<string, string> = {
      "weather in london":
        "The weather in London is currently cloudy with a temperature of 15C.",
      "capital of france": "The capital of France is Paris.",
      "population of earth":
        "The estimated population of Earth is around 8 billion people.",
      "tallest mountain":
        "Mount Everest is the tallest mountain above sea level.",
    };
    const normalizedQuery: string = query.toLowerCase();
    const defaultResult: string = `Simulated search result for '${query}': No specific information found, but the topic seems interesting.`;
    const result: string = simulatedResults[normalizedQuery] ?? defaultResult;

    console.log(`--- TOOL RESULT: ${result} ---`);
    return result;
  },
  {
    name: "search_information",
    description:
      "Return simulated factual information for a few preset queries.",
    schema: z.string().describe("The search query text."),
  },
);

const tools: Array<StructuredToolInterface> = [searchInformationTool];

async function runAgentWithTool(
  executor: AgentExecutorLike,
  query: string,
): Promise<void> {
  console.log(`\n--- Running Agent with Query: '${query}' ---`);

  try {
    const response: AgentResponse = await executor.invoke({
      input: query,
    });
    const outputText: string = String(response.output ?? "");

    console.log("\n--- Final Agent Response ---");
    console.log(outputText);
  } catch (error: unknown) {
    const errorMessage: string =
      error instanceof Error ? error.message : String(error);
    console.log(`\nAn error occurred during agent execution: ${errorMessage}`);
  }
}

async function main(): Promise<void> {
  const config: RuntimeConfig = resolveRuntimeConfig(process.env);
  const llm: ChatOpenAI = new ChatOpenAI({
    apiKey: config.apiKey,
    model: config.model,
    temperature: 0,
    configuration: {
      baseURL: config.baseUrl,
    },
  });
  const prompt: ReturnType<typeof ChatPromptTemplate.fromMessages> =
    ChatPromptTemplate.fromMessages([
      ["system", "You are a helpful assistant."],
      ["human", "{input}"],
      ["placeholder", "{agent_scratchpad}"],
    ]);
  const agent = await createToolCallingAgent({
    llm,
    tools,
    prompt,
  });
  const executor: AgentExecutorLike = new AgentExecutor({
    agent,
    tools,
    verbose: true,
  });

  console.log(`Language model initialized: ${config.model}`);

  await runAgentWithTool(executor, "What is the capital of France?");
  await runAgentWithTool(executor, "What's the weather like in London?");
  await runAgentWithTool(executor, "Tell me something about dogs.");
}

const cliPath: string | undefined = process.argv[1];
if (cliPath) {
  const cliUrl: string = pathToFileURL(resolve(process.cwd(), cliPath)).href;
  if (cliUrl === import.meta.url) {
    void main().catch((error: unknown): void => {
      const errorMessage: string =
        error instanceof Error ? error.message : String(error);
      const message: string = `Error initializing language model: ${errorMessage}`;
      console.error(message);
      process.exitCode = 1;
    });
  }
}
